File: trie/LC2416.py
# LC2316 https://leetcode.cn/problems/sum-of-prefix-scores-of-strings/
from collections import defaultdict
from typing import *


class Node:
    __slots__ = ['next', 'score', 'ids']

    def __init__(self):
        self.next = defaultdict(Node)
        self.score = 0
        self.ids = []


class Solution:
    def sumPrefixScores(self, words: List[str]) -> List[int]:
        root = Node()
        for i, word in enumerate(words):
            cur = root
            for ch in word:
                cur = cur.next[ch]
                cur.score += 1
            cur.ids.append(i)
        ans = [0] * len(words)

        def dfs(node, acc):
            if node is None:
                return
            acc += node.score
            for id in node.ids:
                ans[id] = acc
            for child in node.next.values():
                dfs(child, acc)
        dfs(root, 0)
        return ans

File: trie/LC2416.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    struct Node {
        Node* next[26];
        int score = 0;
        vector<int> ids;
    };
    
    vector<int> ans;
    void dfs(Node* node, int sum){
        if(node==nullptr) return;
        sum += node->score;
        for (int id: node->ids){
            ans[id] = sum;
        }
        for (Node* child: node->next) {
            dfs(child, sum);
        }
    }

    vector<int> sumPrefixScores(vector<string>& words) {
        Node* root = new Node();
        int n = words.size();
        for(int i=0;i<n;++i){
            Node* cur = root;
            for(char ch: words[i]) {
                if (cur->next[ch-'a'] == nullptr) {
                    cur->next[ch-'a'] = new Node();
                }
                cur = cur->next[ch-'a'];
                cur->score++;
            }
            cur->ids.push_back(i);
        }
        ans.resize(n);
        dfs(root, 0);
        return ans;
    }
};